tests/scripting_api/test_window.py

```diff
@@ -0,0 +1,162 @@
+])
+def test_window_center_window(title, xrandr_raw_output, matchClass):
+    window = create_window()
+    xrandr = MagicMock(return_value = xrandr_raw_output)
+    with patch("autokey.scripting.window.Window.set_property") as set_prop:
+        with patch("autokey.scripting.window.Window.resize_move") as resize_move:
+            with patch("autokey.scripting.window.Window._run_xrandr", xrandr) as randr:
+                window.center_window(title)
+                # called with only checks the last time the method was called
+                # set_prop.assert_called_with(title, "remove", "maximized_vert", matchClass=matchClass)
+                # resize_move.assert_called_with(title, , , , )
+                set_prop.assert_called_with(title, "remove", "maximized_horz", matchClass=matchClass)
+
+@pytest.mark.parametrize("wmctrl_output, func_output", [
+    (
+        (0, """0x02000007  0 samdesktop Desktop
+0x0200000f  0 samdesktop nemo-desktop"""),
+    [
+        ("0x02000007","0","samdesktop", "Desktop"),
+        ("0x0200000f", "0" ,"samdesktop", "nemo-desktop")
+    ]
+)])
+def test_window_get_window_list(wmctrl_output, func_output):
+    print(wmctrl_output)
+    #Example output of
+    #$ wmctrl -l
+    #0x02000007  0 samdesktop Desktop
+    #0x0200000f  0 samdesktop nemo-desktop
+    #0x02a00003  0 samdesktop Add more functions to window API by sebastiansam55 · Pull Request #471 · autokey/autokey — Mozilla Firefox
+    #0x02a0004d  1 samdesktop (2698) Best of the Worst: Wheel of the Worst #20 - YouTube — Mozilla Firefox
+    #0x04000007  1 samdesktop Bluetooth Devices
+    #0x04400004  0 samdesktop tmux /home/sam
+    #0x06600004  0 samdesktop python3 /home/sam/git/ak-fork/lib
+    #0x0700000a  0 samdesktop Downloads
+    #0x07c00003  0 samdesktop Find Results (autokey)
+    #0x07800004  0 samdesktop wmctrl /home/sam
+    #0x08400014  0 samdesktop ak-fork - [~/git/ak-fork]
+    #0x07200007  0 samdesktop AutoKey
+    ########
+    # Each line consists of 4 items in order
+    #1. hexid - Hex identifier of the window used by other methods
+    #2. desktop - Which "desktop" (currently called workspaces in Ubuntu 20.04)
+    #3. hostname - Hostname of the computer the window is located on (?)
+    #   Not sure exactly what charaters are allowed in the hostname
+    #4. title - Title of the window
+    window = create_window()
+    wmctrl = MagicMock(return_value = wmctrl_output)
+    with patch("autokey.scripting.window.Window._run_wmctrl", wmctrl):
+        assert func_output == window.get_window_list()
+
+
+@pytest.mark.parametrize("title, func_output, get_window_list_output",[
+    ("AutoKey", "0x07200007", [("0x07200007", "0", "samdesktop", "AutoKey"),("0x02000007", "0", "samdesktop", "Desktop")]),
+])
+def test_window_get_window_hex(title, func_output, get_window_list_output):
+    #TODO there is probably some way to not have to write this at the top of every function
+    window = create_window()
```

Comment:
There is a way, but it isn't much difference because you still have to pass it in as an argument. ([Example here](https://github.com/autokey/autokey/blob/29a8b86ecc0ed3e210b1845e2599b94a305aaf03/tests/scripting_api/test_engine.py#L82) [fixture defined here](https://github.com/autokey/autokey/blob/29a8b86ecc0ed3e210b1845e2599b94a305aaf03/tests/engine_helpers.py#L35)) This is fine as a one-liner.

That being said, I'm not an expert so I might have missed a better way to do it. Pytest's docs on fixtures are a little confusing.

Reply:
Might be able to use the example I linked, but make it an [autouse fixture](https://docs.pytest.org/en/6.2.x/fixture.html#autouse-fixtures-fixtures-you-don-t-have-to-request) (`@pytest.fixture(autouse=True)`)